[Orders with Coupons M7]  Add informational tooltips

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10839. 
#10837 should be merged first
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR adds informational tooltips on tapping the `(?)` symbol when adding coupons to an order, or when adding discounts to a product. This PR deals with the UI only, in the following we'll deal with the logic that makes coupons and discounts mutually exclusive.

| Orders | Orders tooltip |
|--------|--------|
| ![_coupons_tooltip](https://github.com/woocommerce/woocommerce-ios/assets/3812076/e9e86077-dbbe-4518-8549-c601774a6a4d) | ![Simulator Screenshot - iPhone 15 - 2023-10-10 at 20 19 42](https://github.com/woocommerce/woocommerce-ios/assets/3812076/1ab53dbe-8205-4d9b-841f-054fa8288374) | 

| Product | Product tooltip |
|--------|--------|
| ![_discounts_tooltip](https://github.com/woocommerce/woocommerce-ios/assets/3812076/e8ab8d3a-17ab-4655-94ff-4cb5660b01d3)  | ![Simulator Screenshot - iPhone 15 - 2023-10-10 at 20 20 02](https://github.com/woocommerce/woocommerce-ios/assets/3812076/9da74d4f-850e-4c7a-8c80-e54746e95542) | 


## Changes:
* Extracts the product discount row into a `@ViewBuilder` variable for cleaning up the main view. We also move all other `@ViewBuilder` vars to a private extension. 
* Adds `shouldShowCouponsInfoTooltip` state which will change when the `(?)` icon is tapped. This will show/hide an overlay with the informational tooltips.

## Testing instructions
* Switch `ordersWithCouponsM6` feature flag to true
* Go to `Orders` > `+` > Observe that:
  * `+ Add Coupon` is disabled. `(?)` is tappable. Tapping on it toggles the informational tooltip. Tapping again hides it.
  * Tap on `+ Add Product` and add a product to the order. Now `+ Add Coupon` is enabled, and the `(?)` icon is hidden.
  * On the product you've added to the order, or a new product added to the order, tap the chevron to expand the collapsable product row.
  * Observe that the `(?)` in `+ Add discount` is tappable, tapping it shows an informational tooltip and tapping again hides it.
  * Adding a discount to the product will hide both the `+ Add discount` and the informational tooltip icon.

iamgabrielma committed:
Add toggeable tooltip on button tap

iamgabrielma committed:
Add toggeable tooltip on discount icon button tap

iamgabrielma committed:
Disable coupon line if order has discounts

iamgabrielma committed:
Disable discount line if order has coupons

iamgabrielma committed:
Remove whitespace

iamgabrielma committed:
revert logic for existing coupons in parent order

iamgabrielma committed:
Hide info icon when coupons are active

iamgabrielma committed:
Move constants and strings to private extension

iamgabrielma committed:
Move discount const & strings to private extension

iamgabrielma committed:
revert logic to disable coupons if there is discount

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/OrderPaymentSection.swift:
```diff
@@ -82,35 +88,36 @@ struct OrderPaymentSection: View {
             }
 
             addCouponRow
+                .zIndex(1)
```
This is an interesting one: My understanding is that you'd need a ZStack in order to apply `.zIndex`, but that [doesn't seem to be the case](https://developer.apple.com/documentation/swiftui/view/zindex(_:)). For some reason I wasn't able to make the tooltip appear properly on top of the other views without transparency (despite setting opacity specifically, or trying different alternatives) and using this property is the only thing that fixed it.

joshheald replied:
Nothing to add I'm afraid. Maybe comment about it?

iamgabrielma replied:
Added internal comment: 5122b2c900be724137b3799a4205bdf57f091678

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:
```diff
@@ -103,6 +75,28 @@ struct CollapsibleProductRowCard: View {
+                .padding()
+                .background {
+                    Color.black
+                        .cornerRadius(Layout.frameCornerRadius)
```
You should (hopefully) be able to use the `.shadow(.drop)` modifier here to add the drop shadow.

Also... there's some fun to be had adding the little arrow at the top. Seems like a prime candidate for some ChatGPT help if you're not sure how to do it (I'm not sure either, but I think it'd be by combining two other shapes. Figma dev mode suggests using an image of the pointer but that seems like a cop out.)

iamgabrielma replied:
It seems that property is only available for iOS16 😞 . I've added an item to the [Polish UI issue](https://github.com/woocommerce/woocommerce-ios/issues/10863) so I can explore other modifiers.

> Also... there's some fun to be had adding the little arrow at the top. Seems like a prime candidate for some ChatGPT help if you're not sure how to do it (I'm not sure either, but I think it'd be by combining two other shapes. Figma dev mode suggests using an image of the pointer but that seems like a cop out.)

This was tricky, but I got a decent solution with c41c005a99bc2644babdb60643fcc52d61983215 and bc15d28b6151e8beaaa2b187cc7c0f8685de0243 . It needs more tweaking in order to make it fully reusable but at the moment works for our case. I plan to revisit this shortly, perhaps using ~~[.tooltip](https://developer.apple.com/documentation/uikit/uicontrol/3750832-tooltip)~~ [.popover](https://developer.apple.com/documentation/swiftui/view/popover(ispresented:attachmentanchor:arrowedge:content:)) which seems to be a better solution for what we want, and includes the pointer by default. This is the current tooltip UI:

| Coupons tooltip | Discounts tooltip |
|--------|--------|
| ![Simulator Screenshot - iPhone 15 - 2023-10-12 at 10 06 55](https://github.com/woocommerce/woocommerce-ios/assets/3812076/b3e304d4-665d-4a4c-9394-b8c2801701f1) | ![Simulator Screenshot - iPhone 15 - 2023-10-11 at 20 07 44](https://github.com/woocommerce/woocommerce-ios/assets/3812076/2c7ac5ad-6368-4918-93c2-eed8d3f7d6a9) |

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:
```diff
@@ -116,6 +110,44 @@ struct CollapsibleProductRowCard: View {
 }
 
+private extension CollapsibleProductRowCard {
+    @ViewBuilder var discountRow: some View {
```
Nice one for extracting this.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:
```diff
@@ -116,6 +110,44 @@ struct CollapsibleProductRowCard: View {
+        }
+        Spacer()
+            .renderedIf(!viewModel.hasDiscount)
+        Button {
+            shouldShowCouponsInfoTooltip.toggle()
```
For turning it _off_ again, I think we need more than just the button. I instinctively just tapped elsewhere on the screen and that just opens other modals

iamgabrielma replied:
Good catch, addressed in 61f28b59f35f213108d9489780dfbd72485ca27d and b8d3b86938bc9ba8fde244be9ab50b100b0deed7. It's still not perfect, but covers most of the tappable screen.

joshheald reviewed (approved):
Added some questions/suggestions. I think we should do them (improve dismissal, give the tooltips more contextual presence by adding the drop shadows and little arrows.)

Once done, and separately from this project, it would be an interesting task to extract the tooltip to be a bit more reusable.

iamgabrielma committed:
Add internal comment about prop usage

iamgabrielma committed:
Dismiss tooltip when tapping within view

iamgabrielma committed:
Rename to make generic

iamgabrielma committed:
Merge branch 'trunk' into issue/10839-order-ux-m7-info-tooltips

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift

iamgabrielma committed:
Restore redacting discounts while is not sync

Added originally on 3bf84cd3dbfefe98a8e68f69835d2de78e8c0393 , was removed by mistake when merging trunk and fixing merge conflict in the latest commit.

iamgabrielma committed:
Add pointer shape to informational tooltip

iamgabrielma committed:
Make TooltipView reusable

iamgabrielma committed:
Fix tooltip padding

iamgabrielma committed:
Make tooltip hidden on tap gesture

iamgabrielma committed:
fix whitespace
